src/mz/src/bin/mz/upgrader.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Command-line version checker.
+
+use mz::{error::Error, ui::OutputFormatter};
+use mz_build_info::build_info;
+use reqwest::redirect::Policy;
+use reqwest::Client;
+use semver::Version;
+
+use std::io::Write;
+use std::path::PathBuf;
+use std::time::SystemTime;
+use std::{env, path::Path};
+use std::{
+    fs::{File, OpenOptions},
+    io::{BufRead, BufReader},
+};
+
+const CACHE_FILE_NAME: &str = ".mz.ver";
+const BINARIES_LATEST_VERSION_URL: &str =
+    "https://binaries.materialize.com/mz-latest-x86_64-unknown-linux-gnu.tar.gz";
+
+pub struct UpgradeChecker {
+    no_color: bool,
+}
+
+impl UpgradeChecker {
+    pub fn new(no_color: bool) -> Self {
+        UpgradeChecker { no_color }
+    }
+
+    /// Writes the current timestmap in the temp file.
+    fn update_cache_file(&self, version: &Version) -> Result<(), Error> {
+        let mut cache_path: PathBuf = dirs::cache_dir().ok_or(Error::CacheDirNotFoundError)?;
```

Comment:
I could be confused, but AIUI, this is literally a cache? It's our memory of the latest version of `mz` that's available. We store it locally to avoid having to hit the internet on every `mz` run to check if an update is available.